docs/pallas/tpu/collectives.rst

```diff
@@ -0,0 +1,452 @@
+.. note::
+   The :code:`emit_pipeline` API is implemented entirely using public Pallas APIs, an illustration of Pallas meta-programming.
+
+Basic Nested Pipeline Example
+******************************
+
+.. code-block:: python
+
+  def existing_matmul_kernel(x_ref, y_ref, z_ref):
+      @pl.when(pl.program_id(2) == 0)
+      def _():
+        z_ref[...] = jnp.zeros(z_ref.shape, jnp.float32)
+
+      z_ref[...] += x_ref[...] @ y_ref[...]
+
+  def parent_matmul_kernel(x_ref, y_ref, z_ref):
+    pltpu.emit_pipeline(
+        existing_matmul_kernel,
+        grid=(4, 4, 4),
+        in_specs=[
+            pl.BlockSpec(lambda i, j, k: (i, k), (128, 128)),
+            pl.BlockSpec(lambda i, j, k: (k, j), (128, 128)),
+        ],
+        out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (128, 128)),
+    )(x_ref, y_ref, z_ref)
+
+  z = pl.pallas_call(
+      parent_matmul_kernel,
+      out_shape=jax.ShapeDtypeStruct((512, 512), jnp.float32),
+      in_specs=[
+          pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+          pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+      ],
+      out_specs=pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+  )(x, y)
+
+Nested Pipeline with Output Accumulation
+*****************************************
+
+.. code-block:: python
+
+  def matmul_kernel(x_ref, y_ref, z_ref):
+
+    def parent_matmul_kernel(should_accumulate_out):
+      pltpu.emit_pipeline(
+          existing_matmul_kernel,
+          grid=(4, 4, 4),
+          in_specs=[
+            pl.BlockSpec(lambda i, j, k: (i, k), (128, 128)),
+            pl.BlockSpec(lambda i, j, k: (k, j), (128, 128)),
+        ],
+        out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (128, 128)),
+        # Prefix-pytree of out_specs.
+        should_accumulate_out=should_accumulate_out,
+      )(x_ref, y_ref, z_ref)
+
+    emit_pipeline(False)
+    emit_pipeline(True)
+
+  z = pl.pallas_call(
+      matmul_kernel,
+      out_shape=jax.ShapeDtypeStruct((512, 512), jnp.float32),
+      in_specs=[
+          pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+          pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+      ],
+      out_specs=pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+  )(x, y)
+
+Fusing Nested Pipelines (Realistic AG Matmul)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To solve the fact that each child pipeline introduces its own bubbles at their start and end, we must add a layer of complexity via prologue and epilogue callbacks. We danced around with more restrictive but simpler ways of expressing this, but ultimately the behavior here will be very specific to the collective pattern being implemented and you might want to prefetch or wait on pipeline arguments differently based on which ones are participating in the collective.
+
+For pipeline fusion to work we also need to allocate things at the parent grid level. :code:`emit_pipeline_with_allocations` makes that simple.
+
+Here is a full realistic example:
```

Comment:
The jump in complexity from the previous toy pipelines that had no communication is very large here. Can we ease into it?